EIPS/eip-2016.md

```diff
@@ -0,0 +1,241 @@
+---
+eip: 2016
+title: Wallet Add Ethereum Chain RPC Method (`wallet_addEthereumChain`)
+author: Erik Marks (@rekmarks), Pedro Gomes (@pedrouid)
+discussions-to: https://github.com/ethereum/EIPs/issues/3086
+status: Draft
+type: Standards Track
+category: Interface
+created: 2020-11-01
+requires: 155, 695
+---
+
+## Simple Summary
+
+An RPC method for adding Ethereum chains to wallet applications.
+
+## Abstract
+
+The `wallet_addEthereumChain` RPC method allows Ethereum applications ("dapps") to suggest chains to be added to the wallet application.
+The caller must specify a chain ID and some chain metadata.
+The wallet application may arbitrarily refuse or accept the request.
+`true` is returned if the chain was added, and an error otherwise.
+
+Important cautions for implementers of this method are included in the [Security Considerations](#security-considerations) section.
+
+## Motivation
+
+Although Mainnet is by far the most important, a dapp may require the user to interact with other Ethereum chains in order to function.
+Such a chain may or may not be supported by the user's wallet application.
+`wallet_addEthereumChain` enables dapps to request any necessary chains to be added to the user's wallet.
+This provides UX benefits for dapps and wallets that want to provide such functionality.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+### `wallet_addEthereumChain`
+
+The method takes a `chainId` and some chain metadata, returning `true` if the chain was added to the wallet, and an error otherwise.
+
+The wallet **MAY** reject the request for any reason. The wallet **MUST** reject the request if the `chainId` is improperly formatted.
+
+> Note that this method makes **no** statement about whether the wallet should change the user's currently selected chain, if the wallet has a concept thereof.
+
+#### Parameters
+
+```typescript
+interface AddEthereumChainParams {
+  chainId: string;
```

Comment:
Why is this a string?  `chainId` is *always* a number I believe, and passing it as a string just makes using this API harder.

Reply:
Found the answer, and I disagree with it:
> It should be submitted as a `0x`-prefixed hexadecimal string per [EIP-695](./eip-695.md), and parse to an integer number.

695 had chainId as a hex string because at the time it was effectively a uint256, which cannot be reliably decoded across all languages if encoded in json as a number.  However, I think that since then chainId has been narrowed to a uint32 (I forget where) so we no longer need to encode it as a string on the wire, and we should avoid encoding things as strings where we can.

Reply:
IMO, that ship has sailed. `eth_chainId` has always returned a hex string, and whatever benefits there are to encoding it as a `number` are overshadowed by the drawbacks of different RPC methods specifying different encodings for the same value. I'd much rather we live with what we have.